src/components/Location.react.js

```diff
@@ -1,39 +1,85 @@
 import {Component, PropTypes} from 'react';
 /* global window:true */
 
+/**
+ * Update and track the current window.location object through the window.history state
+ */
 export default class Location extends Component {
     constructor(props) {
         super(props);
         this.updateLocation = this.updateLocation.bind(this);
     }
 
     updateLocation(props) {
-        const {pathname, setProps} = props;
+        const {
+            hash,
+            href,
+            pathname,
+            refresh,
+            search,
+            setProps
+        } = props;
 
-        /*
-         * if pathname isn't defined, then it wasn't set by the user
-         * and it needs to be equal to the window.location.pathname
-         * this just happens on page load
+        /**
+         * Check if the field exists in props. If the prop with "fieldName" is not defined,
+         * then it was not set by the user and needs to be equal to the value in window.location.
+         * This only happens on page load (since props will no longer be undefined after componentDidMount).
+         *
+         * @param {string} fieldName
+         *  The name of the prop in window.location and in the component's prop
+         *
+         * @returns {boolean}
+         *  Returns true if the prop with fieldName is different and the window state needs to be updated
          */
-        if (!pathname) {
-            setProps({pathname: window.location.pathname});
-        } else if (pathname !== window.location.pathname) {
-            if (props.refresh) {
-                // Refresh the page
-                window.location.pathname = pathname;
-            } else {
-                // Just push the new location into the URL
-                if (setProps) setProps({pathname});
-                window.history.pushState({}, '', pathname);
+        const checkExistsUpdateWindowLocation = (fieldName) => {
+            const propVal = props[fieldName];
+
+            if (!propVal && window.location[fieldName] && setProps) { // Prop is undefined?
+                const update = {};
+                update[fieldName] = window.location[fieldName];
+                setProps(update);
```

Comment:
Instead of updating the props 1 by 1, I think we should collect all of the properties and update them with a single call to `setProps`. Otherwise, I believe that Dash will rerender the app on every call to `setProps`

Reply:
OK. That's what I had originally done, but was running into some issues so broke them up to make the logic clearer in testing. I'll change it back since the code is working now with tests :+1: